designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+for await (const result of eslint.executeOnFiles(patterns)) {
+  if (Math.random() < 0.5) {
+    break // abort linting.
+  }
+}
+```
+
+And it throws an error if you reuse the aborted results.
+
+##### Implementation
+
+<details>
+<summary>A rough sketch of the `executeOnFiles()` method.</summary>
+
+A tiny wrapper of `CLIEngine`.
+
+```js
+class ESLint {
+  executeOnFiles(patterns) {
+    let promise
+    try {
+      const report = this.cliEngine.executeOnFiles(patterns)
+      promise = Promise.resolve(report.results)
+    } catch (error) {
+      promise = Promise.reject(error)
+    }
+
+    return {
+      then(onFulfilled, onRejected) {
+        return promise.then(onFulfilled, onRejected)
+      },
+      async *[Symbol.asyncIterator]() {
+        yield* await promise
+      },
+    }
+  }
+}
+```
+
+</details>
+
+#### § The `executeOnText()` method
+
+This method returns the same type of an object as the `executeOnFiles()` method.
+
+Because the returned object of `CLIEngine#executeOnText()` method is the same type as the `CLIEngine#executeOnFiles()` method. The `ESLint` class inherits that mannar.
+
+<details>
+<summary>Example: Show the result.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+const [result] = await eslint.executeOnText(text, filePath)
+print(result)
+```
+
+</details>
+
+<details>
+<summary>Example: Using along with the `executeOnFiles()` method.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+const report = useStdin
+  ? eslint.executeOnText(text, filePath)
+  : eslint.executeOnFiles(patterns)
+
+for await (const result of report) {
+  print(result)
+}
+```
+
+</details>
+
+#### § The `getFormatter()` method
+
+This method returns a `Promise<Formatter>`. The `Formatter` type is a function `(results: AsyncIterable<LintResult>) => AsyncIterableIterator<string>`. It receives lint results then outputs the formatted text.
+
+This means the `getFormatter()` method wraps the current formatter to adapt the interface.
+
+<details>
+<summary>A rough sketch of the `getFormatter()` method.</summary>
+
+```js
+class ESLint {
+  async getFormatter(name) {
+    const format = this.cliEngine.getFormatter(name)
+
+    // Return the wrapper.
+    return async function* formatter(resultIterator) {
+      // Collect and sort the results.
+      const results = []
+      for await (const result of resultIterator) {
+        results.push(result)
+      }
+      results.sort(ESLint.compareResultsByFilePath)
+
+      // Make `rulesMeta`.
+      const rules = this.cliEngine.getRules()
+      const rulesMeta = getRulesMeta(rules)
+
+      // Format the results with the formatter of the current spec.
+      yield format(results, { rulesMeta })
+    }
```

Comment:
Why is this an async generator when all it ever yields is a single item? Is the intention that it should be able to somehow format the results and yield them while they keep coming in?

If so: Maybe have the example reflect that?

Reply:
For the future.

`getFormatter()` is a public API, so we couldn't support streaming and printing progress in custom parsers without breaking changes.

If the `getFormatter()` returns this adapter, then we can add custom formatter v2 to support streaming without breaking changes in the future.